### ISSUE TITLE
Improve support for serializing typed EntityRef

### ISSUE THREAD

**Issue by leviramsey** [akka/akka#29955]
The documentation for custom serializers describes support for serializing `ActorRef[T]` via `akka.actor.typed.ActorRefResolver`.  There isn't a counterpart for `EntityRef[T]`.  In principle all that could be required is adding the `entityType`/`entityId`/(optionally) `datacenter` to `EntityRef`'s public API: a deserializer taking an `ExtendedActorSystem` could then easily construct an `EntityRef`.

Alternatively, an `EntityRefResolver` could do the same thing without modifying the public API (though multi-DC support would be more complicated, unless the targeted datacenter is directly encoded into `EntityRefImpl`) of `EntityRef`, though it would be adding a class to the public API.

A do-less alternative might update the documentation to stress that `EntityRef`s should not be included in messages (or persistence events/state) which might be serialized, at least by a custom serializer (I do wonder how `EntityRef`s get handled in the Jackson serializer, which I try to avoid because of reflection fears).  By extension, this implies that there's not much point in having `RecipientRef`s in such situations (one could now directly use a `TestProbe` in addition to an `ActorRef`).

**Comment by johanandren** [comment]
I don't think we should provide a built in serializer or utils for that, but I think making the id and _maybe_ the type available from the `EntityRef` sounds ok. That way a user could potentially make their own serializer if they really want, and resolve a local entity ref using `ClusterSharding.entityRefFor` on the remote side.

**Comment by leviramsey** [comment]
Am I correct in suspecting that serializing an `EntityRefImpl` with the Jackson serialization would deserialize it with a `shardRegion` in the serializing `ActorSystem`?  So if an `EntityRef` is serialized as part of a message after deserialization the recipient would have to send a message to the `shardRegion` on the sender (and this would make an `EntityRef` useless after deserialization if the first system it was serialized on is down)?

**Comment by johanandren** [comment]
If reflection based serialization throws in the entire actor system it would probably not be able ot actually serialize that, and definitely not deserialize it. A serializer would have to encode the entity id, the right amount of details from the typekey to look the same up from some form of registry on the receiving side (since the factory lambda is one of the things and that cannot be serialized).

**Comment by johanandren** [comment]
But as I said earlier, I think we should leave that up to users wanting that rather than provide it out of the box.

**Comment by johanandren** [comment]
I may have ben wrong about the need to look up the type key instance, looks like it is maybe just used for presentation in `EntityRefImpl` and `ask` internals.

**Comment by leviramsey** [comment]
For this test case with Jackson JSON serialization:

```
import akka.actor.typed._
import akka.actor.typed.scaladsl._
import akka.cluster.sharding.typed.scaladsl.{ Entity => ShardingEntity, _}
import akka.serialization._

import scala.concurrent.Await
import scala.concurrent.duration.Duration

import java.nio.file._

object test {
  def main(args: Array[String]): Unit = {
    val system = ActorSystem(Behaviors.empty, "main")

    val sharding = ClusterSharding(system)

    sharding.init(ShardingEntity(Entity.TypeKey) { entityContext =>
      Entity(entityContext.entityId, sharding)
    })

    val entityRefFor: String => EntityRef[Entity.Command] = Entity.refFor(sharding)

    val ref1 :: ref2 :: Nil = List("one", "two").map(entityRefFor)

    val serialization = SerializationExtension(system)

    val path = Paths.get("./cmdserialized.bytes")
    val original = Entity.RecordSiblingEntity(ref2)
    val serializerId = serialization.findSerializerFor(original).identifier

    if (Files.notExists(path)) {
      val bytes = serialization.serialize(original).get

      Files.write(path, bytes)
      println(s"Serialized command to $path")
      system.terminate()
    } else {
      val bytes = Files.readAllBytes(path)
      val cmd = serialization.deserialize(bytes, serializerId, Some(classOf[Entity.RecordSiblingEntity])).get

      ref1 ! cmd
    }

    Await.result(system.whenTerminated, Duration.Inf)
  }
}

trait JacksonSerializable

object Entity {
  sealed trait Command extends JacksonSerializable

  case class RecordSiblingEntity(ref: EntityRef[Command]) extends Command

  val TypeKey = EntityTypeKey[Command]("Entity")

  def refFor(sharding: ClusterSharding)(id: String): EntityRef[Command] =
    sharding.entityRefFor(TypeKey, id)

  def apply(id: String, sharding: ClusterSharding): Behavior[Command] =
    Behaviors.receive { (ctx, msg) =>
      msg match {
        case RecordSiblingEntity(ref) =>
          println(s"Recording sibling: $ref")
          ref ! RecordSiblingEntity(refFor(sharding)(id))
          Behaviors.receiveMessage { _ =>
            ctx.system.terminate()
            Behaviors.empty
          }
      }
    }
}
```

The serialized `RecordSiblingEntity` keys on the path of the `shardRegionActor` (after the first run via sbt)

```
{
  "ref": {
    "shardRegion": "akka://main@192.168.1.32:25520/system/sharding/Entity#-69493921",
    "entityId": "two",
    "typeKey": {
      "name": "Entity",
      "messageClassName": "Entity$Command"
    },
    "refPrefix": "Entity-two",
    "terminated": false
  }
}
```

And in the second run:

> Cannot construct instance of `akka.cluster.sharding.typed.scaladsl.EntityRef` (no Creators, like default construct, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information

Deserializing `EntityRef`s seems to require some custom deserialization even if using Jackson, in which case one may as well use custom serialization and eliminate the extraneous fields.

**Comment by johanandren** [comment]
I think that you will need a custom serializer (Akka or a Jackson module) is to be expected since the `EntityRef` is not a pojo but a little bit more complicated structure with an internal impl implementing two public APIs for Java and Scala respectively.
